### DISCUSSION TITLE
Send key pressed event from code

### DISCUSSION THREAD

**Discussion by IngvarX** [AvaloniaUI/Avalonia#5687] [Q&A]
Hello!

I'm trying to send key pressed event from code. I'm trying to do it in a following way:

```csharp
 this.FindAncestorOfType<MainWindow>().RaiseEvent(new KeyEventArgs
            {
                RoutedEvent = InputElement.KeyUpEvent,
                Key = Key.F4,
                KeyModifiers = KeyModifiers.Alt
            });
```

But this event is ignored by app. Anything I'm doing wrong?

**Comment by maxkatz6** [comment]
I believe Source property setting is missed in your code

**Reply by IngvarX** [reply]
Code like this doesn't work as well:

```csharp
his.FindAncestorOfType<MainWindow>().RaiseEvent(new KeyEventArgs
            {
                RoutedEvent = InputElement.KeyDownEvent, //KeyUpEvent
                Key = Key.F4,
                KeyModifiers = KeyModifiers.Alt,
                Source = this.FindAncestorOfType<MainWindow>()
            });
```

Anything specific that I should use as source?

**Reply by maxkatz6** [reply]
It depends where you want to handle this event.
Let's say you have KeyDown event handler in your CustomUserControl, you need to use this CustomUserControl as a source or any inner control inside of it, so it can be raised back to the CustomUserControl.

You can also raise this event on that source directly instead of MainWindow.

**Reply by IngvarX** [reply]
I was able to send keys using this:

```csharp
KeyboardDevice.Instance.ProcessRawEvent(
                new RawKeyEventArgs(
                    KeyboardDevice.Instance, 0,  window,
                    RawKeyEventType.KeyDown, Key.F7, RawInputModifiers.None));
```

but it doesn't work when I call it in headless test in `AfterSetup` method. any ideas?

**Reply by maxkatz6** [reply]
Can you share some code/project how your headless platform is set up?

**Reply by IngvarX** [reply]
https://github.com/IngvarX/Camelot/blob/F7-test/tests/Camelot.Ui.Tests/AvaloniaApp.cs

**Reply by IngvarX** [reply]
Resolved this issue, it happened due to focus
